src/frontend/core/executor/impl/TriangleCountExecutor.cpp

```diff
@@ -34,13 +50,391 @@ TriangleCountExecutor::TriangleCountExecutor(SQLiteDBInterface *db, PerformanceS
+        transfer[it->first] = it->second;
+    }
+    net_loads.clear();
+    net_loads.insert(best_plan.net_loads.begin(), best_plan.net_loads.end());
+    loads.clear();
+    loads.insert(best_plan.loads.begin(), best_plan.loads.end());
+    return best;
+}
+
+static void filter_partitions(std::map<string, std::vector<string>> &partitionMap, SQLiteDBInterface *sqlite,
+                              string graphId) {
+    map<string, string> workers;  // id => "ip:port"
+    const std::vector<vector<pair<string, string>>> &results =
+        sqlite->runSelect("SELECT DISTINCT idworker,ip,server_port FROM worker;");
+    for (int i = 0; i < results.size(); i++) {
+        string workerId = results[i][0].second;
+        string ip = results[i][1].second;
+        string port = results[i][2].second;
+        workers[workerId] = ip + ":" + port;
+    }
+
+    map<string, int> loads;
+    const map<string, string> &cpu_map = Utils::getMetricMap("cpu_usage");
+    for (auto it = workers.begin(); it != workers.end(); it++) {
+        auto workerId = it->first;
+        auto worker = it->second;
+        const auto workerLoadIt = cpu_map.find(worker);
+        if (workerLoadIt != cpu_map.end()) {
+            double load = stod(workerLoadIt->second.c_str());
+            if (load < 0) load = 0;
+            loads[workerId] = (int)(4 * load);
+        } else {
+            loads[workerId] = 0;
+        }
+    }
+
+    std::map<int, std::vector<string>> p_avail;
+    std::set<int> remain;
+    for (auto it = partitionMap.begin(); it != partitionMap.end(); it++) {
+        auto worker = it->first;
+        auto &partitions = it->second;
+        for (auto partitionIt = partitions.begin(); partitionIt != partitions.end(); partitionIt++) {
+            auto partition = stoi(*partitionIt);
+            p_avail[partition].push_back(worker);
+            remain.insert(partition);
+        }
+    }
+    const std::map<int, std::vector<string>> P_AVAIL = p_avail;  // get a copy and make it const
+
+    for (auto loadIt = loads.begin(); loadIt != loads.end(); loadIt++) {
+        if (loadIt->second < 3) continue;
+        auto w = loadIt->first;
+        for (auto it = p_avail.begin(); it != p_avail.end(); it++) {
+            auto &partitionWorkers = it->second;
+            for (auto workerIt = partitionWorkers.begin(); workerIt != partitionWorkers.end();) {
+                if (*workerIt == w) {
+                    partitionWorkers.erase(workerIt);
+                } else {
+                    workerIt++;
+                }
+            }
+        }
+    }
+
+    std::map<int, string> alloc;
+    int unallocated = alloc_plan(alloc, remain, p_avail, loads);
+    if (unallocated > 0) {
+        triangleCount_logger.info(to_string(unallocated) + " partitions remaining after alloc_plan");
+        auto copying = reallocate_parts(alloc, remain, P_AVAIL);
+        scale_up(loads, workers, copying.size());
+        triangleCount_logger.info("Scale up completed");
```

Comment:
We need more context details in this log message. Do we also log when scale up started?

Reply:
The previous log (3 lines before this log message) gives the time when scale up requested.